### ISSUE TITLE
[Plugin] Software Bill of Materials (SBOM)

### ISSUE THREAD

**Issue by regicsolutions** [backstage/backstage#13630]
Suggesting a plugin to display a Software Bill of Materials (SBOM) Composition Report. [CycloneDX](https://cyclonedx.org/use-cases/) seems to be a commonly used standard across a number of [tools](https://cyclonedx.org/tool-center/) that supports both [JSON](https://cyclonedx.org/docs/1.4/json/#) and [XML](https://cyclonedx.org/docs/1.4/xml/) schemas.

[CycloneDX](https://cyclonedx.org/) is the SBOM format created by the Open Web Application Security Project (OWASP) and was built with cybersecurity in mind. The largest benefit of CycloneDX is the ability to include vulnerability information in the SBOM or link to a separate [Vulnerability Exploitability Exchange (VEX)](https://cyclonedx.org/capabilities/vex/) which lets you indicate false positives alongside component vulnerability information.

Unfortunately i don’t have time to contribute right now but wanted to throw the idea out there.

Thinking the plugin will work similar to the code-coverage plugin which has a [front-end](https://github.com/backstage/backstage/tree/master/plugins/code-coverage) and [backend-plugin](https://github.com/backstage/backstage/tree/master/plugins/code-coverage-backend) and can support multiple SBOM Providers listed [here](https://cyclonedx.org/tool-center/)

`metadata:
  annotations:
    backstage.io/sbom: enabled`

A [tab](https://backstage.io/storybook/?path=/story/navigation-tabs--default) called SBOM to be shown at the component, system and even resource level with sub tabs with the SBOM provider name i.e Sonatype, Prisma or any tool that supports the Cyclone specification standard.

**Plugin Use Case:** I am using [Sonatype Nexus Lifecycle](https://help.sonatype.com/iqserver/analysis/cyclonedx-application-analysis) to evaluate my application for open-source vulnerabilities which produces a SBOM composition report with a full list of my applications packages and dependencies along with any associated vulnerabilities. I also use another tool called [Prisma Cloud](https://www.paloaltonetworks.com/prisma/cloud/cloud-code-security) that also supports the Cyclone specification that scans my applications container for operating system level dependencies and vulnerabilities and would like to have a central view to both SBOMs at build time and runtime.    

Unfortunately i don’t have time to contribute right now but wanted to throw the idea out there and get feedback from the community!

**Comment by testn** [comment]
any one wants to take a stab at this?

**Comment by regicsolutions** [comment]
> any one wants to take a stab at this?

Hi @testn this is currently open for grabs, would you like to be assigned? 

Unfortunately, I don't have the time at the moment, just wanted to throw the idea out there to see if there was any interest/feedback. 

Docker now actually has a [sbom plugin](https://docs.docker.com/engine/sbom/) that generates a CycloneDX SBOM in JSON. Felt this plugin would come in handy to visualize the SBOM report at the component level.

**Comment by jkowalleck** [comment]
Hi there.
I am the author of https://github.com/CycloneDX/cyclonedx-javascript-library
Feel free to reach out to me if you need something regarding the library, or just want to chat about CDX in general.

If you want to use custom properties in CycloneDX, then you might want to register your namespace here: https://github.com/CycloneDX/cyclonedx-property-taxonomy/

**Comment by DJDANNY123** [comment]
I did some work on a proof-of-concept CycloneDX plugin back in early 2022 before much tooling existed for CycloneDX, I'll see what I can contribute back and if it's still relevant but my approach/ideas on a plugin can be summarised below:

SBOMs are living documents that should be updated regularly, so they should be linked to rather than captured within a component yaml. 

They could be stored within a repository such as https://github.com/CycloneDX/cyclonedx-bom-repo-server or potentially within VCS alongside the code, For my initial attempt I just got it working with harvesting from Git using the default URLReader and relying on the established auth mechanism.

I added a tab on each software component page to summarise some of the information contained within an SBOM, e.g. loop through the dependencies and list the number of each license within the SBOM etc, but with the CycloneDX library mentioned there could be a more informative/browsable UI. 

Some inspiration for the UI could be taken from plugins like [https://github.com/backstage/backstage/blob/master/plugins/fossa/README.md](FOSSA) or [https://github.com/snyk-tech-services/backstage-plugin-snyk/blob/main/README.md](Snyk) as similar information may be contained in the SBOM.

**Comment by regicsolutions** [comment]
> I did some work on a proof-of-concept CycloneDX plugin back in early 2022 before much tooling existed for CycloneDX, I'll see what I can contribute back and if it's still relevant but my approach/ideas on a plugin can be summarised below:
> 
> 
> 
> SBOMs are living documents that should be updated regularly, so they should be linked to rather than captured within a component yaml. 
> 
> 
> 
> They could be stored within a repository such as https://github.com/CycloneDX/cyclonedx-bom-repo-server or potentially within VCS alongside the code, For my initial attempt I just got it working with harvesting from Git using the default URLReader and relying on the established auth mechanism.
> 
> 
> 
> I added a tab on each software component page to summarise some of the information contained within an SBOM, e.g. loop through the dependencies and list the number of each license within the SBOM etc, but with the CycloneDX library mentioned there could be a more informative/browsable UI. 
> 
> 
> 
> Some inspiration for the UI could be taken from plugins like [https://github.com/backstage/backstage/blob/master/plugins/fossa/README.md](FOSSA) or [https://github.com/snyk-tech-services/backstage-plugin-snyk/blob/main/README.md](Snyk) as similar information may be contained in the SBOM.

@DJDANNY123 Would love to see what you built so far, I agree that sboms are living/always changing documents and pulling from git is definitely a great start, given sboms are generated during build time, was thinking it would great to be able to post the sbom report similar how the [code-coverage](https://github.com/backstage/backstage/tree/master/plugins/code-coverage-backend) accepts a coverage report.

**Comment by DJDANNY123** [comment]
Had a meeting yesterday to progress getting permission from third party to open source this. Waiting for formal confirmation from them to release it.

That being said I think I would want to rewrite some of the work I did to use the library above to process the SBOM rather than manually decode it using the JSON schemas etc. 

I was also returning the entire SBOM to the front-end which is probably not the approach I'd like to take, as it would be good to process that on the backend and then return the extracted information on the frontend, or get data from a report as well as a raw SBOM, do you have an example of any existing SBOM report data structures?

**Comment by regicsolutions** [comment]
Agreed, [here are some SBOM sample reports](https://github.com/CycloneDX/bom-examples/tree/master/SBOM) 

The [cyclonedx-javascript-library](https://github.com/CycloneDX/cyclonedx-javascript-library/tree/main)](https://github.com/CycloneDX/cyclonedx-javascript-library) also has a example of how to serialize a SBOM to JSON [here](https://github.com/CycloneDX/cyclonedx-javascript-library/blob/main/examples/node-typescript/example.mjs/src/example.ts)

Let me know if you would like to be assigned to this one!

**Comment by DJDANNY123** [comment]
Ah maybe I misunderstood the terminology of 'report', I was viewing there would be a raw SBOM which is based on the CycloneDX standard, or a 'report' which contains a more minimal representation of the information contained within this, such as a list of licenses, vulnerabilities etc.

I was thinking the backend component would retrieve a full CycloneDX SBOM and process this in the backend to a more minimal JSON structure or 'Report' that provides information extracted from the SBOM (e.g. metadata, list of contained licenses, vulnerabilities etc.) to be displayed in Backstage in addition to a link to the raw SBOM.

I don't know if there are other services that produce/expose the processed information from an SBOM or a format around this which we could also use, but maybe thats beyond the scope of a CycloneDX plugin.

**Comment by jkowalleck** [comment]
> Agreed, [here are some SBOM sample reports](https://github.com/CycloneDX/bom-examples/tree/master/SBOM) @jkowalleck would you happen to have any samples using the latest [1.4 spec](https://cyclonedx.org/docs/1.4/json/)?

There are more on the official documentation and website: <https://cyclonedx.org/use-cases/>
Examples might be nice, but you should study the spec - <https://github.com/CycloneDX/specification/tree/master/schema>


> I don't know if there are other services that produce/expose the processed information from an SBOM [...]

You might connect with the CycloneDX community and check what others were already doing.
Or ask them for SBOMs they've created. 

* join the slack: <https://cyclonedx.org/slack/invite>
* join the group: https://groups.io/g/CycloneDX

**Comment by jkowalleck** [comment]
if it might help you, there is now a list of CycloneDX guides : https://cyclonedx.org/guides/

**Comment by elifarley** [comment]
Maybe SPDX support is also important. If SPDX support is somehow easier to get, I'd be happy to have it first.

**Comment by elifarley** [comment]
Regarding the storage location for the SBOM (in CycloneDX, SPDX or any other format), it could make more sense to store it as an OCI artifact alongside the container images (_[ORAS](https://oras.land/) can be used to attach supply-chain artifacts to container images_).

Indexing all SBOMS would also be great for searching.
